{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ca667087-4f5b-4711-9c2f-954f29a942f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from pathlib import Path\n",
    "from collections import defaultdict\n",
    "\n",
    "import requests\n",
    "import pandas as pd\n",
    "\n",
    "DATA_DIR = Path('../association/onek1k_website_data/')\n",
    "OUTPUT_DIR = Path('../')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0f7947d-f82f-4e53-b18b-6b4eeabd9031",
   "metadata": {},
   "source": [
    "# API Clients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "de0b9dd0-bda6-4336-8dab-a0e9b2a79b1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "class ApiClient:\n",
    "    \"\"\"\n",
    "    Base API client providing common cache operations.\n",
    "    \"\"\"\n",
    "    def __init__(self, base_url, connection=None, cache=None):\n",
    "        self.base_url = base_url\n",
    "        self.connection = connection or requests\n",
    "        self.cache = {}\n",
    "\n",
    "        if isinstance(cache, (Path, str)):\n",
    "            self.load_cache(file_path=cache)\n",
    "        \n",
    "    def save_cache(self, file_path):\n",
    "        with open(file_path, 'wt') as file:\n",
    "            json.dump(self.cache, file)\n",
    "        return self\n",
    "            \n",
    "    def load_cache(self, file_path):\n",
    "        with open(file_path, 'rt') as file:\n",
    "            self.cache = json.load(file)\n",
    "        return self"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc15512b-5711-44ba-bb6e-6113896cc4d9",
   "metadata": {},
   "source": [
    "## MyVariant\n",
    "Simple client for [MyVariant](https://myvariant.info/v1/api) service to retrieve dbSNP record information such as consequence etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "22373a71-0919-4da9-aaad-ca5e501e951e",
   "metadata": {},
   "outputs": [],
   "source": [
    "class MyVariantClient(ApiClient):\n",
    "    \"\"\"\n",
    "    Simple client for https://myvariant.info/v1/api\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, base_url='http://myvariant.info/v1/api', connection=None, cache=None):\n",
    "        super.__init__(self, base_url=base_url, connection=connection, cache=cache)\n",
    "        \n",
    "    def fetch_rsid(self, identifier):\n",
    "        if identifier in self.cache:\n",
    "            return self.cache[identifier]\n",
    "        \n",
    "        result = self.connection.get(f\"{self.base_url}/variant/{identifier}\").json()\n",
    "        self.cache[identifier] = result\n",
    "        return result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "150a5f48-d4a9-48f8-9a32-3296bc0ca2ed",
   "metadata": {},
   "source": [
    "## HGNC\n",
    "Simple client for [HGNC rest API](https://www.genenames.org/help/rest/) to retrieve gene information such as names, aliases, transcripts etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b15f22a2-6db3-4904-bcc5-81b8678f6e58",
   "metadata": {},
   "outputs": [],
   "source": [
    "class HgncClient(ApiClient):   \n",
    "    \"\"\"\n",
    "    Simple client for https://www.genenames.org/help/rest/\n",
    "    \"\"\"\n",
    "        \n",
    "    def __init__(self, base_url='http://rest.genenames.org', connection=None, cache=None):\n",
    "        super().__init__(base_url=base_url, connection=connection, cache=cache)\n",
    "\n",
    "    def fetch(self, identifier, fields=('hgnc_id')):\n",
    "        if isinstance(fields, str):\n",
    "            fields = (fields,)\n",
    "            \n",
    "        if identifier in self.cache.keys():\n",
    "            field_mapping = self.cache[identifier]\n",
    "            requested_fields = set(fields) & set(field_mapping.keys())\n",
    "            for field in requested_fields:\n",
    "                if field_mapping.get(field, None):\n",
    "                    return field_mapping[field]\n",
    "            if\n",
    "        else:\n",
    "            self.cache[identifier] = {}\n",
    "        \n",
    "        for field in fields:\n",
    "            response = self.connection.get(\n",
    "                f\"{self.base_url}/fetch/{field}/{identifier}\",\n",
    "                headers={'Accept': 'application/json'}\n",
    "            ).json()\n",
    "\n",
    "            records = response['response'].get('docs', [])\n",
    "            if not records:\n",
    "                records = None\n",
    "                \n",
    "            self.cache[identifier][field] = records\n",
    "            \n",
    "            if records:\n",
    "                break\n",
    "        \n",
    "        return records\n",
    "    \n",
    "    def fetch_many(self, identifiers, fields=('hgnc_id')):\n",
    "        records = {}\n",
    "        for identifier in identifiers:\n",
    "            records[identifier] = None\n",
    "            result = self.fetch(identifier, field=field)\n",
    "            for field in fields:\n",
    "                \n",
    "                if result is not None:\n",
    "                    records[identifier] = result\n",
    "                    break\n",
    "        return records"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3eb1a94-1e56-4e26-ab35-76cdc902e489",
   "metadata": {},
   "source": [
    "# Generate Gene search file\n",
    "Generate `gene_search_terms.json.txt` file where each line consists of a valid `JSON` list element. Each of these elements is a tuple that looks like `[gene_name, [search_terms]]`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "383a1280-7366-4b93-8468-cee922e24f36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching metadata for 'ARSA'\n",
      "Fetching metadata for 'C22orf34'\n",
      "Fetching metadata for 'CHCHD10'\n",
      "Fetching metadata for 'CRYBB2'\n",
      "Fetching metadata for 'CTA-29F11.1'\n",
      "Fetching metadata for 'APOBEC3C'\n",
      "Fetching metadata for 'ARFGAP3'\n",
      "Fetching metadata for 'BCR'\n",
      "Fetching metadata for 'CHCHD10'\n",
      "Fetching metadata for 'CRYBB2'\n",
      "Fetching metadata for 'AC000068.5'\n",
      "Fetching metadata for 'APOBEC3G'\n",
      "Fetching metadata for 'APOL2'\n",
      "Fetching metadata for 'ARSA'\n",
      "Fetching metadata for 'BCR'\n",
      "Fetching metadata for 'ADRBK2'\n",
      "Fetching metadata for 'APOBEC3G'\n",
      "Fetching metadata for 'APOBEC3H'\n",
      "Fetching metadata for 'APOL6'\n",
      "Fetching metadata for 'ARSA'\n",
      "Fetching metadata for 'AC002472.13'\n",
      "Fetching metadata for 'APOBEC3G'\n",
      "Fetching metadata for 'APOBEC3H'\n",
      "Fetching metadata for 'ARSA'\n",
      "Fetching metadata for 'ASPHD2'\n",
      "Fetching metadata for 'ARSA'\n",
      "Fetching metadata for 'CTA-29F11.1'\n",
      "Fetching metadata for 'DDT'\n",
      "Fetching metadata for 'FAM118A'\n",
      "Fetching metadata for 'GGT1'\n",
      "Fetching metadata for 'DDT'\n",
      "Fetching metadata for 'FAM118A'\n",
      "Fetching metadata for 'IGLL1'\n",
      "Fetching metadata for 'LGALS2'\n",
      "Fetching metadata for 'MIF'\n",
      "Fetching metadata for 'ADORA2A-AS1'\n",
      "Fetching metadata for 'APOBEC3A'\n",
      "Fetching metadata for 'APOBEC3B'\n",
      "Fetching metadata for 'CBX6'\n",
      "Fetching metadata for 'CDC42EP1'\n",
      "Fetching metadata for 'APOBEC3A'\n",
      "Fetching metadata for 'ARSA'\n",
      "Fetching metadata for 'BCR'\n",
      "Fetching metadata for 'DDT'\n",
      "Fetching metadata for 'FAM118A'\n",
      "Fetching metadata for 'ARSA'\n",
      "Fetching metadata for 'CTA-29F11.1'\n"
     ]
    }
   ],
   "source": [
    "gene_search_terms = defaultdict(list)\n",
    "gene_metadata = {}\n",
    "\n",
    "# ---- Manually insert info not available online\n",
    "gene_metadata['AC007308.6'] = [{\n",
    "    'ensembl_gene_id': 'ENSG00000234252',\n",
    "    'symbol': 'AC007308.6',\n",
    "    'name': 'AC007308.6 (Clone-based (Vega))',\n",
    "    'location': '22qxx.x'\n",
    "}]\n",
    "gene_metadata['AC002472.13'] = [{\n",
    "    'ensembl_gene_id': 'ENSG00000187905',\n",
    "    'symbol': 'AC002472.13',\n",
    "    'name': 'Leucine-rich repeat-containing protein LOC400891',\n",
    "    'location': '22qxx.x'\n",
    "}]\n",
    "gene_metadata['CTA-29F11.1'] = [{\n",
    "    'ensembl_gene_id': 'CTA-29F11.1',\n",
    "    'symbol': 'CTA-29F11.1',\n",
    "    'name': 'lncRNA',\n",
    "    'location': '0qxx.x'\n",
    "}]\n",
    "gene_metadata['AC000068.5'] = [{\n",
    "    'ensembl_gene_id': 'ENSG00000185065',\n",
    "    'symbol': 'AC000068.5',\n",
    "    'name': 'AC000068.5 (Clone-based (Vega) gene)',\n",
    "    'location': '19qxx.x'\n",
    "}]\n",
    "# ----------------------------------------------\n",
    "cache_path = OUTPUT_DIR / 'hgnc_cache.json'\n",
    "if cache_path.exists():\n",
    "    hgnc_client = HgncClient(cache=cache_path)\n",
    "else:\n",
    "    hgnc_client = HgncClient(cache=None)\n",
    "\n",
    "cell_labels = sorted([\n",
    "    expression_file.name.split('_')[0].strip() \n",
    "    for expression_file in DATA_DIR.glob('*_eQTLs.tsv')\n",
    "])\n",
    "    \n",
    "for label in cell_labels:\n",
    "    df = pd.read_csv(DATA_DIR / f\"{label}_eQTLs.tsv\", header=0, delimiter='\\t')\n",
    "    for gene, group in df.groupby('GENE'):\n",
    "        print(f\"Fetching metadata for '{gene}'\")\n",
    "        metadata = hgnc_client.fetch(\n",
    "            identifier=gene,\n",
    "            fields=('symbol', 'prev_symbol', 'alias_symbol', 'vega_id', 'uniprot_ids')\n",
    "        )\n",
    "        \n",
    "        if metadata:\n",
    "            gene_metadata[gene] = metadata\n",
    "            gene_search_terms[gene].append(metadata[0]['ensembl_gene_id'])\n",
    "            gene_search_terms[gene].append(metadata[0]['hgnc_id'])\n",
    "            gene_search_terms[gene].append(metadata[0]['symbol'])\n",
    "            gene_search_terms[gene].append(metadata[0]['name'])\n",
    "        \n",
    "        gene_search_terms[gene].append(label)\n",
    "        gene_search_terms[gene].append(gene)\n",
    "        gene_search_terms[gene] += list(group['SNP'])\n",
    "        gene_search_terms[gene] += list(f\"chr{chrom}\" for chrom in group['CHR'])\n",
    "        \n",
    "hgnc_client.save_cache(OUTPUT_DIR / 'hgnc_cache.json')\n",
    "\n",
    "with open(OUTPUT_DIR / 'gene_search_terms.json.txt', 'wt') as file:\n",
    "    for gene, search_terms in gene_search_terms.items():\n",
    "        file.write(json.dumps([gene, list(set(search_terms))]))\n",
    "        file.write('\\n')\n",
    "        \n",
    "with open(OUTPUT_DIR / 'gene_metadata.json', 'wt') as file:\n",
    "    json.dump(gene_metadata, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82642882-7bf8-47d9-8c07-f367df6a4092",
   "metadata": {},
   "source": [
    "# Site metadata file\n",
    "This is a metadata file containing configuration options for the TOB browser. These are used by the framework to automate table rendering and file serving on the client and server side.\n",
    "\n",
    "Framework *requires* the keys `reference_genome`, `gene_result_analysis_groups` and `gene_group_result_field_names`. Other fields are TOB specific."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4f4a846a-002d-4048-9146-9234e1a22d16",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_readable_labels = {\n",
    "    \"BimmNaive\": \"Bimm Naive\",\n",
    "    \"Bmem\": \"B Memory\",\n",
    "    \"CD4all\": \"CD4 All\",\n",
    "    \"CD8all\": \"CD8 All\",\n",
    "    \"CD8eff\": \"CD8 Eff\",\n",
    "    \"CD8unknown\": \"CD8 Unknown\",\n",
    "    \"DC\": \"DC\",\n",
    "    \"MonoC\": \"Mono C\",\n",
    "    \"MonoNC\": \"Mono CN\",\n",
    "    \"NKact\": \"NK Act\",\n",
    "    \"NKmat\": \"NK Mat\",\n",
    "    \"Plasma\": \"Plasma\"\n",
    "}\n",
    "\n",
    "metadata = {\n",
    "    \"datasets\": {\n",
    "        \"tob\": {\n",
    "            \"reference_genome\": \"GRCh37\",\n",
    "            \"gene_result_analysis_groups\": [\"All\"],\n",
    "            \"gene_group_result_field_names\": cell_labels,\n",
    "            \"gene_results_table_headings\": {\n",
    "                l: human_readable_labels[l] for l in cell_labels\n",
    "            },\n",
    "            \"gene_symbols\": sorted(gene_search_terms.keys())\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "with open(OUTPUT_DIR / 'metadata.json', 'wt') as file:\n",
    "    json.dump(metadata, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f288bcd7-c190-44c7-a9d2-821727b566ff",
   "metadata": {},
   "source": [
    "# Generate gene results file\n",
    "Generates the gene results `JSON` file that the exome browser framework uses to render a gene results table when requesting the `/results` URL."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b95d31ee-3963-4d51-a24a-aa72b7b08c81",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'C22orf34'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-ee26dc972310>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m         \u001b[0mcell_label_results\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'GENE'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mgene\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m     \u001b[0mmetadata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgene_metadata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m     record = [\n\u001b[1;32m     10\u001b[0m         \u001b[0mmetadata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'ensembl_gene_id'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'C22orf34'"
     ]
    }
   ],
   "source": [
    "gene_results = {\"results\": []}\n",
    "for gene in gene_search_terms.keys():\n",
    "    cell_label_results = []\n",
    "    for cell_label in cell_labels:\n",
    "        df = pd.read_csv(DATA_DIR / f\"{cell_label}_eQTLs.tsv\", header=0, delimiter='\\t')\n",
    "        cell_label_results.append(df[df['GENE'] == gene].shape[0])\n",
    "        \n",
    "    metadata = gene_metadata[gene][0]\n",
    "    record = [\n",
    "        metadata.get('ensembl_gene_id'),\n",
    "        gene,\n",
    "        metadata.get('name'),\n",
    "        int(metadata.get('location').split('q')[0]),\n",
    "        1,\n",
    "        [cell_label_results]\n",
    "    ]\n",
    "    \n",
    "    gene_results['results'].append(record)\n",
    "\n",
    "with open(OUTPUT_DIR / 'results'/ 'tob.json', 'wt') as file:\n",
    "    json.dump(gene_results, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "199ca335-7ef0-47d5-b37b-12b44d834314",
   "metadata": {},
   "source": [
    "# Process eQTL association files\n",
    "Process the raw association files into a `JSON` mapping indexed by cell type and then further by gene name. Values are a list of SNP mappings containing p-values, log10 p-values, dbSNP identifiers and other analysis data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "503bf0c7-0458-4ead-8ee5-19491a5e6e0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_associations = defaultdict(lambda: defaultdict(list))\n",
    "\n",
    "for label in cell_labels:\n",
    "    df = pd.read_csv(DATA_DIR / f\"{label}_eQTLs.tsv\", header=0, delimiter='\\t')\n",
    "    for gene, group in df.groupby('GENE'):\n",
    "        records = list(group.to_dict(orient='records'))\n",
    "        for r in records:\n",
    "            r['ID'] = f\"{r['CHR']}-{r['BP']}-{r['A1']}-{r['A2']}\"\n",
    "            r['CONSEQUENCE'] = 'TODO'\n",
    "            r['HGVSP'] = 'TODO'\n",
    "            r['HGVSC'] = 'TODO'\n",
    "        snp_associations[label][gene] = records\n",
    "        \n",
    "with open(OUTPUT_DIR / 'snp_association.json', 'wt') as file:\n",
    "    json.dump(snp_associations, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "167a39ed-7b64-4dab-b565-a6b12ad675ff",
   "metadata": {},
   "source": [
    "# Process expression files\n",
    "Each file ending in `*_residual expression.tsv` contains expression data for a given gene listed in the columns. Each file lists different genes so we need to concatenate each dataframe initialising missing columns to `0`. This file will be served for UMAP construction. Saved in `"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe4446b4-060a-4f01-bad8-28aeaf9bebd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = {\n",
    "    label: pd.read_csv(DATA_DIR / f\"{label}_residual_expressions.tsv\", header=0, delimiter='\\t')\n",
    "    for label in cell_labels\n",
    "}\n",
    "\n",
    "genes = list(set(g for df in dfs.values() for g in df.columns ))\n",
    "for gene in genes:\n",
    "    for cell_label, df in dfs.items():\n",
    "        df['cell_label'] = cell_label\n",
    "        if gene not in df.columns:\n",
    "            df[gene] = 0\n",
    "            \n",
    "expression = pd.concat(dfs.values(), axis=0, ignore_index=True)\n",
    "expression = expression[[\"cell_label\"] + sorted(genes)]\n",
    "expression.to_csv(OUTPUT_DIR / 'results' / 'cell_label_expression.csv', index=False, sep=\",\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
